### ISSUE TITLE
Large system clock drift can prevent nodes from joining gossip

### ISSUE THREAD

**Issue by t-nelson** [solana-labs/solana#29198]
#### Problem
gossip filters received push and pull messages by +/-30s and +/-15s respectively.  if a node's system clock has drifted sufficiently, the rest of the cluster will ignore their messages and they the messages from the rest of the cluster.  at startup, this prevents the crds table from populating at all and the effected node never discovers peers.

#### Proposed Solution
add a sanity check for system time at startup

some things to avoid
* block time. it is affected by slot times as well as outages, so not sufficiently reliable for this task
* web2 time oracles. this is a blockchain.

instead we can probably measure the rate of gossip message drops due to falling outside the acceptable wallclock window.  if it's sufficiently high after timing out joining gossip, log a message suggesting that the clock be set and exit

**Comment by steviez** [comment]
Random note - In the event of cluster restarts, there is also some filtering done based on wall clock time here while calculating what % of the cluster is up.
https://github.com/solana-labs/solana/blob/dcfb73f664a33c09ac75b1771bd09d978e6d44c6/core/src/validator.rs#L1946-L1955

The hypothetical check proposed would probably happen before the snippet I just provided, but just mentioning this other are for completeness.

**Comment by bw-solana** [comment]
I'm observing that we aren't seeing too many error metrics incremented as a result of inaccurate system clock. I'm seeing CRDS updates get filtered due to the age window, but we're only receiving dozens of these. What seems more pronounced is the lack of responses to our pull requests.

Does this seem like a good signal to key off of? I.e. If the response rate to our gossip pull requests is extremely low, and we fail to find RPC candidates to download snapshot from, print the error message and exit

**Comment by t-nelson** [comment]
> but we're only receiving dozens of these

who's "we"?  i wouldn't expect a large number of nodes to be affected because the situation typically needs to be resolved for the validator to join the cluster reliably at all

**Comment by bw-solana** [comment]
> > but we're only receiving dozens of these
> 
> who's "we"? i wouldn't expect a large number of nodes to be affected because the situation typically needs to be resolved for the validator to join the cluster reliably at all

Me and my friendly test validator 👯 . After spinning up Gossip service, I'm seeing a handful of Pull Requests trickle in, but we filter them out due to wallclock falling outside the expected window.

I'm concerned that the volume here is low and it's hard to distinguish between a system clock setting issue on this validator vs. its peer (or a malicious peer).

> 
> > Does this seem like a good signal to key off of?
> 
> maybe. i'm concerned that that'd be noisy because there are plenty of other reasons for a pull request to be dropped on the other end

That's true, but I'm seeing 0 pull requests receive a response out of >20k requests generated. I'm thinking this dramatically low response rate + failure to get a snapshot should give strong signal

**Comment by bw-solana** [comment]
> test validator's clock is intentionally busted?

oui

**Comment by bw-solana** [comment]
> is the busted node also dumping pushes from its peers? i'd expect that to be the reason it can't find a snapshot candidate

I'm actually seeing zero push messages coming in.  A few pruning, ping, and pong messages. A few more pull requests. Zero pull responses or push messages

**Comment by t-nelson** [comment]
oh! _that_ might be sufficient criteria.  probably not _zero_ but some exceptionally low rate

@behzadnouri wdyt? i know you find the notion of the CRDS table being in any notion of steady state being a dubious one

**Comment by behzadnouri** [comment]
yeah relying on gossip messages to identify clock skew would probably be risky and not reliable.
e.g. it is subject to network partitions, packets being dropped, gossip acting up, etc
Also those wallclock filters are subject to change, which will make it harder to roll out those changes if it also impacts others nodes perceive of their clock skew. In general I think if we don't tie two different things (gossip processing vs clock skew) together and add dependency between them would be better.

Can't we just bake in some (perhaps optional) ntp sync/check into validator code?
 
I know some folks use chrony and apparently jito is running a chrony server
https://chrony.tuxfamily.org/

**Comment by t-nelson** [comment]
yeah the idea here isn't so much to fix it for the operator, but to raise that the clock being off might be why the node isn't behaving as it ought to be.  i think some false positives would be more acceptable than maintaining some partial NTP implementation for the same purpose.  maybe just need to get some more data so we can characterize the false positive rate?

**Comment by bw-solana** [comment]
yeah, I wouldn't want to tie any logic between these two. Error message suggesting the possibility is the furthest I'd want to go.

On a semi-related note, should we add a timeout to `get_rpc_nodes` around `get_rpc_peers`? Seems like we can return None forever

**Comment by t-nelson** [comment]
> yeah, I wouldn't want to tie any logic between these two. Error message suggesting the possibility is the furthest I'd want to go.

:100: 

> On a semi-related note, should we add a timeout to `get_rpc_nodes` around `get_rpc_peers`? Seems like we can return None forever

i think the "haven't found a snapshot in N minutes" thing will exit the process implicitly in this case

**Comment by bw-solana** [comment]
> i think the "haven't found a snapshot in N minutes" thing will exit the process implicitly in this case

Have a node that's been spinning in that loop for 10+ minutes. I'm not seeing anything that would get us out of it when `cluster_info.all_rpc_peers()` returns nothing

**Comment by behzadnouri** [comment]
One problem with using gossip (or perhaps any other internal component) is that most likely it will behave very hectic during network restarts.
It is also a risk that it may introduce further instability when the network becomes unstable or when certain number of nodes go delinquent/offline; i.e. making things worse right at the worst time.

> i think some false positives would be more acceptable than maintaining some partial NTP implementation for the same purpose.

Likely there are already crates for this.

**Comment by bw-solana** [comment]
First attempt at validating system clock against external server time in the event we cannot find any peers for bootstrap: #30278

**Comment by bw-solana** [comment]
Is the guidance provided in #30277 after failing to find RPC peers sufficient, or is there more we want to do here?

I think #30277 should be sufficient for the case where system clock is way off, but if the delta is right around the CRDS filtering range, we could still limp along and thus never provide guidance message. Do we care?

**Comment by t-nelson** [comment]
let's not care for now and see if the support request decline sufficiently with #30277

**Comment by steviez** [comment]
Interestingly, I observed a node this morning in [validator-support](https://discord.com/channels/428295358100013066/560174212967432193/1110581647537090560) that:
- Had a > 4 minute drift in their system clock (node was behind)
- Was still freezing banks / observing roots at same velocity as cluster

I figured out system clock because the node was computing the same hashes and observing roots but seemingly before the cluster. So, I figured the timestamps on the metrics submissions must be wrong (node with slow clock is purple trace, other two were good nodes):
![image](https://github.com/solana-labs/solana/assets/5400107/29427042-40b5-4703-bb9c-6d663dd7fe4d)

Mentioning it here because the previous occurrences of a bad clock have often resulted in the node being unable to participate in the cluster altogether. This node was a staked node, so maybe it's position in turbine was already "established", or maybe it was able to use the `contact-info.bin` from a previous run and that info was accurate enough?
